### DISCUSSION TITLE
Queries regarding distributed external cache (Redis) utilizing multiple Ocelot nodes

### DISCUSSION THREAD

**Discussion by SheruGaur** [ThreeMammals/Ocelot#2144] [Q&A] [answered]
Dear @raman-m  , I would appreciate your assistance with an issue regarding the configuration of an external Redis cache across multiple Ocelot instances that utilize the same backend Redis server and identical route configurations in each instance.

Below, I have outlined my cache route policy in relation to the external Redis cache.

 #### Redis Configuration code in program.cs
```C#
services.AddOcelot()
 .AddCacheManager(x => x.WithDictionaryHandle())
 .AddCacheManager(x =>
 {
     var redisSettings = configuration.GetSection("RedisSettings");
     string redisIp = redisSettings["Ip"];
     int redisPort = int.Parse(redisSettings["Port"]);
     string redisPassword = redisSettings["Password"]; // Retrieve the password
     x.WithRedisConfiguration("redis",
             config =>
             {
                 config.WithAllowAdmin()
                 .WithDatabase(0)
                 .WithEndpoint(redisIp, redisPort)
                 .WithPassword(redisPassword); // Apply the password
             })
     .WithJsonSerializer()
     .WithRedisCacheHandle("redis");
 })
```
#### Route configuration in each Ocelot API gateway instances
```json
{
  "UpstreamPathTemplate": "/api/test/Employee/GetEmployeeDetails/{EmpCode}",
  "UpstreamHttpMethod": [ "Get" ],
  "DownstreamPathTemplate": "/test/api/Employee/GetEmployeeDetails/{EmpCode}",
  "DownstreamScheme": "https",
  "DownstreamHttpMethod": "Get",
  "DownstreamHostAndPorts": [
    {
      "Host": "testdomain.com",
      "Port": 443
    }
  ],
  "FileCacheOptions": {
    "TtlSeconds": 180,
    "Region": "testregion"
  },
  "Key": "testkey1"
}
```

The following are the observation points for which I require clarification:

1. After configuring Ocelot with a cache that has a Time-to-Live (TTL) of 180 seconds and specifying the Redis as external cache (as described above), I have observed that each Ocelot instance generates its own set of default active cache key-value pairs. 

3. I would like to understand why the identical route configurations result in different cache key-value pairs being generated, rather than a consistent cache key-value pair across different instance of Ocelot nodes. Is this expected?

Please could you suggest what do we need to do differently.

Thanks,
Sheru Gaur

**Comment by raman-m** [comment]
> 1. After configuring Ocelot with a cache that has a Time-to-Live (TTL) of 180 seconds and specifying the Redis as external cache (as described above), I have observed that **each Ocelot instance generates its own set of default active cache key-value pairs**.

OK, could you please show examples of these keys? And what’s the difference?

Also, please add `"EnableContentHashing": false` explicitly, but it should work implicitly for sure.

**Reply by raman-m** [reply]
>2. I would like to understand why the identical route configurations result in different cache key-value pairs being generated, rather than a consistent cache key-value pair across different instance of Ocelot nodes. Is this expected

This behavior is unexpected. Let's investigate the root cause together. It likely stems from [body caching](https://ocelot.readthedocs.io/en/latest/features/caching.html#enablecontenthashing-option). You can disable it [globally](https://ocelot.readthedocs.io/en/latest/features/caching.html#global-configuration) by setting `"EnableContentHashing": false`. 

If this doesn't resolve the issue, then we might be dealing with an Ocelot bug. Please ensure that you distribute the identical configuration across all Ocelot instances using Docker image versioning. Which technique do you employ?

> Ocelot nodes

Sorry? What did you mean actually? Kubernetes nodes?

**Comment by raman-m** [comment, marked as answer]
>  #### Redis Configuration code in program.cs
> ```C#
> services.AddOcelot()
>  .AddCacheManager(x => x.WithDictionaryHandle())
>  .AddCacheManager(x =>
>  {
>      var redisSettings = configuration.GetSection("RedisSettings");
>      // ...
>  })
> ```

Why did you set up twice? You should call the method only once!

### P.S.
I advise against a hybrid setup such as [CacheManager](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20CacheManager.Core&type=code) + Redis. CacheManager should only be used for memory caching. We have been using the [CacheManager.Core package](https://www.nuget.org/packages/CacheManager.Core/) since 2019, but it appears to be deprecated, with the last version being [2.0.0-beta-1629](https://www.nuget.org/packages/CacheManager.Core/2.0.0-beta-1629), released on November 15, 2018. Additionally, the most recent repository release is [1.2.0](https://github.com/MichaCo/CacheManager/releases/tag/1.2.0) , dated December 6, 2018, which is quite outdated.

### P.P.S.
@SheruGaur, please implement [Your Own Caching](https://ocelot.readthedocs.io/en/latest/features/caching.html#your-own-caching).
For distributed caching, you must utilize the [IDistributedCache](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.caching.distributed.idistributedcache?view=net-8.0) interface and substitute `IOcelotCache<T>` in DI. As Ocelot currently lacks support for [distributed caching](https://learn.microsoft.com/en-us/aspnet/core/performance/caching/distributed?view=aspnetcore-8.0), we may consider elevating this discussion to a feature request. A similar problem is discussed in #2098.

### TODO
As a result, our team plans to discontinue support for the [Ocelot.Cache.CacheManager](https://www.nuget.org/packages/Ocelot.Cache.CacheManager) package soon and will update the documentation accordingly.
@RaynaldM @ggnaegi  FYI
